test/jdk/java/util/stream/GatherersMapConcurrentTest.java

```diff
@@ -0,0 +1,277 @@
+        }
+
+        fail("This should not be reached");
+    }
+
+    @ParameterizedTest
+    @MethodSource("small_atleast3_configurations")
+    public void cancelsStartedTasksIfShortCircuited(Config config) {
+        final var streamSize = config.streamSize();
+
+        assertTrue(streamSize > 2, "This test case won't work with tiny streams!");
+
+        final var tasksToCancel = streamSize - 2;
+        final var firstReady = new CountDownLatch(1);
+        final var lastDone = new CountDownLatch(1);
+        final var tasksCancelled = new CountDownLatch(tasksToCancel);
+
+        final var tasksWaiting = new Semaphore(0);
+
+        final var result =
+                config.stream().gather(
+                    Gatherers.mapConcurrent(streamSize, i -> {
+                        switch (i) {
+                            case 1 -> {
+                                firstReady.countDown();
+                                try { lastDone.await(); }
+                                catch (InterruptedException ie) {
+                                    fail("Unexpected!");
+                                }
+                            }
+
+                            case Integer n when n == streamSize -> {
+                                try { firstReady.await(); }
+                                catch (InterruptedException ie) {
+                                    fail("Unexpected!");
+                                }
+                                while(tasksWaiting.getQueueLength() < tasksToCancel) {
+                                    try {
+                                        Thread.sleep(10);
+                                    } catch (InterruptedException ie) {
+                                        // Ignore
+                                    }
+                                }
+                                lastDone.countDown();
+                            }
+
+                            default -> {
+                                try {
+                                    tasksWaiting.acquire();
+                                } catch (InterruptedException ie) {
+                                    tasksCancelled.countDown(); // used to ensure that they all were interrupted
+                                }
+                            }
+                        }
+
+                        return i;
+                    })
+            )
+            .gather(Gatherer.of((unused, state, downstream) -> downstream.push(state) && false)) // emulate limit(1)
+            .toList();
+        assertEquals(List.of(1), result);
+        try {
+            tasksCancelled.await();
+        } catch (InterruptedException ie) {
+            fail("Unexpected");
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("configurations")
+    public void behavesAsExpected(Config config) {
+        for (var concurrency : List.of(1, 2, 3, 10, 1000)) {
+            final var expectedResult = config.stream()
```

Comment:
I assume the loop is here because you can't have both a MethodSource and ValueSource. I think I'd probably create a new method source for this, e.g.

```
    static Stream<Arguments> configutatonAndConcurrency() {
        return IntStream.of(1, 2, 3, 10, 1000)
                .mapToObj(i -> i)
                .flatMap(i -> configurations().map(c -> new Arguments(c, i)));
    }
```

The advantage is that you'll have the parameter in the .jtr in the event of a failure.

Reply:
@AlanBateman Yeah, it's an unfortunate limitation of JUnit. Ideally I'd like to have an annotation per argument and being able to specify that I'd like the framework to execute it as a cartesian product.

Do you want me to make this change as a part of this PR or does it make more sense to get this one integrated first?

Reply:
> Do you want me to make this change as a part of this PR or does it make more sense to get this one integrated first?

I think it's okay to leave it to another time, I assume these tests be updated a few times before the feature becomes permanent.